d2/src/main/java/com/linkedin/d2/balancer/servers/ZooKeeperAnnouncer.java

```diff
@@ -296,14 +344,15 @@ public void onError(Throwable e)
         else
         {
           //continue to mark up to the regular cluster
+          _regularClusterMarkUpStartAt = System.currentTimeMillis();
```

Comment:
good to know. Will update, thanks.

Reply:
Actually the time here is not to measure elapsed time but to be used as time stamp in the events (update intent timestamp, and write complete timestamp), which nanoTime doesn't makes sense since it "represents nanoseconds since some fixed but arbitrary origin time (perhaps in the future, **so values may be negative**)."

The time in ZooKeeperEphemeralStore is for measuring request duration, but the duration needs to be in milliseconds as defined in the event. So if I used nanoTime, I will still need to convert it back to milliseconds before emitting the event, which will lose the precision anyway.

I will keep using milliseconds for these reasons.

Reply:
For distributed measurements across different hosts, we probably care less about time to go back for "a bit", yet if we measure duration on the same host, we should use stable clock.

Reply:
After a second thought, actually doing a Nano -> Milli conversion can still be more precise than directly using milliseconds for the calculation since some system may use tens of milliseconds under-the-hood. Changed to use nanoTime. Thanks.